File: minefield.go
package main

import (
	"math/rand"
	"time"
)

type MineField struct {
	width      int
	height     int
	totalMines int
	minesFound int
	gameOver   bool
	gameWon    bool
	field      [][]Cell
}

func NewMineField(width, height, mines int) *MineField {
	mf := &MineField{
		width:      width,
		height:     height,
		totalMines: mines,
		minesFound: 0,
		gameOver:   false,
		gameWon:    false,
		field:      make([][]Cell, height),
	}
	
	// Инициализация поля безопасными клетками
	for i := 0; i < height; i++ {
		mf.field[i] = make([]Cell, width)
		for j := 0; j < width; j++ {
			mf.field[i][j] = NewSafeCell()
		}
	}
	
	return mf
}

func (mf *MineField) IsValidCoord(x, y int) bool {
	return x >= 0 && x < mf.height && y >= 0 && y < mf.width
}

func (mf *MineField) CountAdjacentMines(x, y int) int {
	count := 0
	for dx := -1; dx <= 1; dx++ {
		for dy := -1; dy <= 1; dy++ {
			if dx == 0 && dy == 0 {
				continue
			}
			nx := x + dx
			ny := y + dy
			if mf.IsValidCoord(nx, ny) && mf.field[nx][ny].HasMine() {
				count++
			}
		}
	}
	return count
}

func (mf *MineField) GenerateMines(firstX, firstY int) {
	minesPlaced := 0
	rand.Seed(time.Now().UnixNano())
	
	for minesPlaced < mf.totalMines {
		x := rand.Intn(mf.height)
		y := rand.Intn(mf.width)
		
		// Не размещаем мину на первой открытой клетке и вокруг неё
		tooClose := false
		for dx := -1; dx <= 1; dx++ {
			for dy := -1; dy <= 1; dy++ {
				if x+dx == firstX && y+dy == firstY {
					tooClose = true
					break
				}
			}
			if tooClose {
				break
			}
		}
		
		if !tooClose && !mf.field[x][y].HasMine() {
			mf.field[x][y] = NewMineCell()
			minesPlaced++
		}
	}
	
	// Подсчет соседних мин для всех клеток
	for i := 0; i < mf.height; i++ {
		for j := 0; j < mf.width; j++ {
			count := mf.CountAdjacentMines(i, j)
			mf.field[i][j].SetAdjacentMines(count)
		}
	}
}

func (mf *MineField) OpenEmptyArea(x, y int) {
	// Используем стек для итеративного открытия
	stack := [][2]int{{x, y}}
	
	for len(stack) > 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		cx, cy := current[0], current[1]
		
		if !mf.IsValidCoord(cx, cy) || mf.field[cx][cy].GetIsOpen() {
			continue
		}
		
		mf.field[cx][cy].SetOpen(true)
		
		// Если клетка пустая, добавляем ее соседей
		if mf.field[cx][cy].GetAdjacentMines() == 0 && !mf.field[cx][cy].HasMine() {
			for dx := -1; dx <= 1; dx++ {
				for dy := -1; dy <= 1; dy++ {
					if dx == 0 && dy == 0 {
						continue
					}
					nx := cx + dx
					ny := cy + dy
					if mf.IsValidCoord(nx, ny) && !mf.field[nx][ny].GetIsOpen() &&
						!mf.field[nx][ny].GetIsFlagged() {
						stack = append(stack, [2]int{nx, ny})
					}
				}
			}
		}
	}
}

func (mf *MineField) OpenCell(x, y int, firstMove bool) bool {
	if !mf.IsValidCoord(x, y) || mf.field[x][y].GetIsOpen() ||
		mf.field[x][y].GetIsFlagged() {
		return true // Игра продолжается
	}
	
	if firstMove {
		mf.GenerateMines(x, y)
	}
	
	if mf.field[x][y].HasMine() {
		mf.field[x][y].SetOpen(true)
		mf.gameOver = true
		return false // Игра окончена - поражение
	}
	
	mf.OpenEmptyArea(x, y)
	mf.CheckWin()
	return !mf.gameOver
}

func (mf *MineField) ToggleFlag(x, y int) {
	if !mf.IsValidCoord(x, y) || mf.field[x][y].GetIsOpen() {
		return
	}
	
	wasFlagged := mf.field[x][y].GetIsFlagged()
	mf.field[x][y].ToggleFlag()
	
	// Обновляем счетчик найденных мин
	if mf.field[x][y].GetIsFlagged() && !wasFlagged && mf.field[x][y].HasMine() {
		mf.minesFound++
	} else if !mf.field[x][y].GetIsFlagged() && wasFlagged && mf.field[x][y].HasMine() {
		mf.minesFound--
	}
	
	mf.CheckWin()
}

func (mf *MineField) CheckWin() {
	// Победа если все мины помечены или все безопасные клетки открыты
	openedSafeCells := 0
	totalSafeCells := mf.width*mf.height - mf.totalMines
	
	for i := 0; i < mf.height; i++ {
		for j := 0; j < mf.width; j++ {
			if !mf.field[i][j].HasMine() && mf.field[i][j].GetIsOpen() {
				openedSafeCells++
			}
		}
	}
	
	if openedSafeCells == totalSafeCells {
		mf.gameWon = true
		mf.gameOver = true
	}
}

func (mf *MineField) IsGameOver() bool {
	return mf.gameOver
}

func (mf *MineField) IsGameWon() bool {
	return mf.gameWon
}

func (mf *MineField) GetMinesFound() int {
	return mf.minesFound
}

func (mf *MineField) GetTotalMines() int {
	return mf.totalMines
}

func (mf *MineField) GetWidth() int {
	return mf.width
}

func (mf *MineField) GetHeight() int {
	return mf.height
}

func (mf *MineField) GetCell(x, y int) Cell {
	if mf.IsValidCoord(x, y) {
		return mf.field[x][y]
	}
	return nil
}

File: MineField.cpp
#include "MineField.h"
#include "MineCell.h"
#include "SafeCell.h"
#include <iostream>
#include <cstdlib>
#include <ctime>

using namespace std;

bool MineField::isValidCoord(int x, int y) const {
    return x >= 0 && x < height && y >= 0 && y < width;
}

int MineField::countAdjacentMines(int x, int y) const {
    int count = 0;
    for (int dx = -1; dx <= 1; dx++) {
        for (int dy = -1; dy <= 1; dy++) {
            if (dx == 0 && dy == 0) continue;
            int nx = x + dx;
            int ny = y + dy;
            if (isValidCoord(nx, ny) && field[nx][ny]->hasMine()) {
                count++;
            }
        }
    }
    return count;
}

void MineField::openEmptyArea(int x, int y) {
    if (!isValidCoord(x, y) || field[x][y]->getIsOpen()) {
        return;
    }
    
    field[x][y]->setOpen(true);
    
    if (field[x][y]->getAdjacentMines() == 0 && !field[x][y]->hasMine()) {
        for (int dx = -1; dx <= 1; dx++) {
            for (int dy = -1; dy <= 1; dy++) {
                if (dx == 0 && dy == 0) continue;
                openEmptyArea(x + dx, y + dy);
            }
        }
    }
}

MineField::MineField(int w, int h, int mines) 
    : width(w), height(h), totalMines(mines), minesFound(0), 
      gameOver(false), gameWon(false) {
    
    // Инициализация поля безопасными клетками
    field.resize(height);
    for (int i = 0; i < height; i++) {
        field[i].resize(width);
        for (int j = 0; j < width; j++) {
            field[i][j] = new SafeCell();
        }
    }
}

MineField::~MineField() {
    for (int i = 0; i < height; i++) {
        for (int j = 0; j < width; j++) {
            delete field[i][j];
        }
    }
}

void MineField::generateMines(int firstX, int firstY) {
    int minesPlaced = 0;
    srand(time(0));
    
    while (minesPlaced < totalMines) {
        int x = rand() % height;
        int y = rand() % width;
        
        // Не размещаем мину на первой открытой клетке и вокруг неё
        bool tooClose = false;
        for (int dx = -1; dx <= 1; dx++) {
            for (int dy = -1; dy <= 1; dy++) {
                if (x + dx == firstX && y + dy == firstY) {
                    tooClose = true;
                    break;
                }
            }
            if (tooClose) break;
        }
        
        if (!tooClose && !field[x][y]->hasMine()) {
            delete field[x][y];
            field[x][y] = new MineCell();
            minesPlaced++;
        }
    }
    
    // Подсчет соседних мин для всех клеток
    for (int i = 0; i < height; i++) {
        for (int j = 0; j < width; j++) {
            int count = countAdjacentMines(i, j);
            field[i][j]->setAdjacentMines(count);
        }
    }
}

bool MineField::openCell(int x, int y, bool firstMove) {
    if (!isValidCoord(x, y) || field[x][y]->getIsOpen() || 
        field[x][y]->getIsFlagged()) {
        return true;  // Игра продолжается
    }
    
    if (firstMove) {
        generateMines(x, y);
    }
    
    if (field[x][y]->hasMine()) {
        field[x][y]->setOpen(true);
        gameOver = true;
        return false;  // Игра окончена - поражение
    }
    
    openEmptyArea(x, y);
    checkWin();
    return !gameOver;
}

void MineField::toggleFlag(int x, int y) {
    if (!isValidCoord(x, y) || field[x][y]->getIsOpen()) {
        return;
    }
    
    bool wasFlagged = field[x][y]->getIsFlagged();
    field[x][y]->toggleFlag();
    
    // Обновляем счетчик найденных мин
    if (field[x][y]->getIsFlagged() && !wasFlagged && field[x][y]->hasMine()) {
        minesFound++;
    } else if (!field[x][y]->getIsFlagged() && wasFlagged && field[x][y]->hasMine()) {
        minesFound--;
    }
    
    checkWin();
}

void MineField::checkWin() {
    // Победа если все мины помечены или все безопасные клетки открыты
    int openedSafeCells = 0;
    int totalSafeCells = width * height - totalMines;
    
    for (int i = 0; i < height; i++) {
        for (int j = 0; j < width; j++) {
            if (!field[i][j]->hasMine() && field[i][j]->getIsOpen()) {
                openedSafeCells++;
            }
        }
    }
    
    if (openedSafeCells == totalSafeCells) {
        gameWon = true;
        gameOver = true;
    }
}

void MineField::display(bool revealAll) const {
    cout << "\n   ";
    for (int j = 0; j < width; j++) {
        cout << j % 10 << " ";
    }
    cout << "\n";
    
    for (int i = 0; i < height; i++) {
        cout << i % 10 << "  ";
        for (int j = 0; j < width; j++) {
            if (revealAll && field[i][j]->hasMine()) {
                cout << "X ";
            } else {
                cout << field[i][j]->getDisplayChar() << " ";
            }
        }
        cout << "\n";
    }
    
    cout << "\nМин всего: " << totalMines;
    cout << " | Найдено: " << minesFound << "\n";
}

bool MineField::isGameOver() const { 
    return gameOver; 
}

bool MineField::isGameWon() const { 
    return gameWon; 
}

int MineField::getMinesFound() const { 
    return minesFound; 
}

int MineField::getTotalMines() const { 
    return totalMines; 
}
